### DISCUSSION TITLE
Reactive Power Configuration or Standby Mode setup by API

### DISCUSSION THREAD

**Discussion by jovana** [WillCodeForCats/solaredge-modbus-multi#296] [Q&A]
Hi, 

Does someone know if it is possible to use the API from solaredge to switch of the inverter?
I want to know because if the energy price is negative, I want to switch off solar production.
I have found two options to do this, but I can not figure out if this is available in the API. 

On page 6 we can set the produced power limit to 0%:
https://knowledge-center.solaredge.com/sites/kc/files/application_note_power_control_configuration.pdf

Or we can enable the standby mode using the setApp: https://www.youtube.com/watch?v=zVD5wDNNSMU

Many thanks for helping!
Jo

**Comment by mpredfearn** [comment]
If you have the "Enable Site Limit Control" configuration option enabled, you should be able to use the site limit control https://github.com/WillCodeForCats/solaredge-modbus-multi/blob/main/custom_components/solaredge_modbus_multi/number.py#L361 to adjust this.
![Screenshot_2023-04-19_14-33-02](https://user-images.githubusercontent.com/12936891/233092277-1f2a2c06-4ac5-4558-a7c4-a14e68e1f728.png)
The controls are as described on page 15 of this document https://github.com/WillCodeForCats/solaredge-modbus-multi/blob/main/doc/Power-Control-Open-Protocol-for-SolarEdge-Inverters.pdf

**Reply by jovana** [reply]
Hi, many thx for your fast reply! 
Okay probably this is what I need. I will dive into this.
I did not enable this setting, jet. Because I read something to use this only if you totally understand 😃.

One more question about this. Has al these settings and updates been parsed using the Modbus connection? 

thx!

**Reply by mpredfearn** [reply]
@jovana yeah, because some peoples inverters are configured by their installers with an export limit to comply with local regulations. Using these settings *potentially* allows you to bypass those controls. But if you have no such restrictions and all you want to to is limit export under conditions like a negative export price, then hopefully this mechanism should work for you.

Sorry I didn't get your last question?

**Reply by WillCodeForCats** [reply]
Also keep in mind that even after enabling controls in the integration, SolarEdge doesn't always make them available over modbus for all inverters. For more info on that issue see the Not All Inverters Support Battery Data or Controls over Modbus section of [Known Issues on the wiki](https://github.com/WillCodeForCats/solaredge-modbus-multi/wiki/Known-Issues)

**Reply by jovana** [reply]
> @jovana yeah, because some peoples inverters are configured by their installers with an export limit to comply with local regulations. Using these settings _potentially_ allows you to bypass those controls. But if you have no such restrictions and all you want to to is limit export under conditions like a negative export price, then hopefully this mechanism should work for you.
> 
> Sorry I didn't get your last question?

Thx! 

My last question, it's about changing settings without any authentication. Reading data using Modbus, I understand, but changing settings feels a bit un-secure. But unfortunately, it's not uncommon in the IoT market.

**Reply by mpredfearn** [reply]
@jovana ah yes, there is no authentication. I suppose since you have to manually enable the modbus server in the inverter, and it only listens on the local network, there's little a bad actor could do unless they were already connected to your network, but in that case you've probably got bigger problems.

**Reply by WillCodeForCats** [reply]
> My last question, it's about changing settings without any authentication. Reading data using Modbus, I understand, but changing settings feels a bit un-secure. But unfortunately, it's not uncommon in the IoT market.

Modbus is simply a protocol from the late 1970's that predates modern ideas about security. There was no IoT market or even commercial internet access at the time. All of your modbus devices would have been on a serial bus that's physically isolated.

**Comment by jtonk** [comment]
Hi, I tried:
Commissioning page, select Power Control → Energy Manager → Limit Control→ Control Mode → Production Control (I can't select export control because I don't have a meter that reports the in/export to the inverter) After setting this I restarted HA, but I don't get the extra controls. Am I doing something wrong?

**Reply by WillCodeForCats** [reply]
That has nothing to do with integration configuration.

Read:
* https://github.com/WillCodeForCats/solaredge-modbus-multi/wiki/Power-Control-Options
* https://github.com/WillCodeForCats/solaredge-modbus-multi/wiki/Configuration#integration-configuration-options

**Reply by WillCodeForCats** [reply]
Also keep in mind that even after enabling controls in the integration, SolarEdge doesn't always make them available over modbus for all inverters. For more info on that issue see the Not All Inverters Support Battery Data or Controls over Modbus section of [Known Issues on the wiki](https://github.com/WillCodeForCats/solaredge-modbus-multi/wiki/Known-Issues)

**Reply by jtonk** [reply]
thanks for the info. I was reading here https://knowledge-center.solaredge.com/sites/kc/files/feed-in_limitation_application_note.pdf that I need a external meter... So this is not going to work for me now.'
Will look into the other items before I install a meter. Any recommendation on meters?
I have tried building my own P1 port (Dutch meter provided by grid provider) to modbus adapter with an esp32, but that didn't work out (yet)....

**Reply by WillCodeForCats** [reply]
You can still use a production site limit without a meter.

**Reply by jtonk** [reply]
not sure how, but problem is solved!

**Comment by jovana** [comment]
After a few weeks of trial and error, I have found (for our home situation) the best setup.
I have dynamic energy prices, so I want to have HA create policies on what to do with PV power and energy storage.

- If prices are negative
  - Disable the PV production.
  - Full charge the battery from the grid
- If prices are positive:
  - If price_level is cheap or very_cheap: 
    - Charge from PV power (before exporting to the grid)
    - Disable discharge (because energy needs to save if the price is going up)
  - If price_level is normal, expensive, or very_expensive:
    - Maximize self-consumption

Below is a visual representation of the automation created in HA.    

![battery - PV policies](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/645456/15ea0cc9-a419-43d5-8f1e-2eee56a8ae74)

**Reply by deqw** [reply]
After playing with these settings I was not able to prevent the inverter to use solar power for in-house use. With today's pricing, with several hours at minus 40 cent per kWh, you want to maximise power from the grid, so not using solar power at all.

With the above settings (Price < 0), while having Storage Charge Limit = 0 W, there is 5 kW going from solar to the battery or to the house. My question is: how to prevent getting solar power into the system at all? (without pulling the cable from the inverter ;o)
![IMG_7289](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/65849349/7f6027dc-d690-44b5-a98f-8dd278f41283)
![IMG_7290](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/65849349/d4ad6a6e-2358-495c-bca3-134a1454004a)

**Reply by jovana** [reply]
Can I ask what kind of inverter and battery you have? Do you have a 3 phase inverter?

**Reply by deqw** [reply]
The inverter is a SE10K-RWB48 (3-phase). The battery is a StorEdge 9.2kWh SolarEdge Home Battery (48V). I also have the SolarEdge Home EV Charger and the Import/Export meter.  Software and maybe the firmware too **_isn't_** already at the level that all works neat together, I noticed.

**Comment by jovana** [comment]
<img width="329" alt="Scherm­afbeelding 2023-07-02 om 13 54 52" src="https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/645456/16a7c96c-3bfb-4e3b-a36d-d14c1595bc6a">

These are the settings in HA. Can you please check you have this also setup?

**Reply by deqw** [reply]
Yes got that settings. See attached. Can set the Storage Charge Limit to whatever it keeps charging exactly what is coming from the solar panels.
![Screenshot 2023-07-02 at 13 57 19](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/65849349/bc34e3e6-2c39-4281-a947-085204516f74)

**Reply by jovana** [reply]
Okay, looks good. Do you have the battery and PV panels on the same inverter?
In my setup, I have separated inverters. 

What happens if you change "Limit Control Mode" to Disabled?

**Reply by deqw** [reply]
I have just one inverter. When I change "Limit Control Mode" to Disabled then solar power is imported again which results in lower grid import.

**Comment by jovana** [comment]
<img width="464" alt="Scherm­afbeelding 2023-07-02 om 13 56 43" src="https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/645456/e4baa8dc-b267-4217-b090-edb81f21aa5f">
It should disabled the PV power production.

**Reply by deqw** [reply]
Yes that is how it should be.

Edit: oh, battery is full now and no solar reported anymore which is different from before.
![IMG_7291](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/65849349/d33e5398-4768-441f-acad-37b30e3b7d67)

**Reply by jovana** [reply]
Okay so it seems like the setting "Storage Command Mode" we need an option "Charge From Grid Only". Not sure if this exists, @WillCodeForCats Seth, do you know if there is an option to fully disable PV but charge the battery from grid only?

**Reply by jovana** [reply]
Lekker 14kw, levert je bijna 6 euro op als je dit een uurtje volhoud! 🤣

**Reply by deqw** [reply]
Thanks for follow up!

Yep car is charging now until about 16:40 and some washing is done. What you say: looks like "Charge From Grid Only" is needed here.

**Reply by WillCodeForCats** [reply]
No, there isn't a charge from grid only option in the specs that I can see.

**Reply by deqw** [reply]
@WillCodeForCats See my reply below: can StorageRemoteCtrl_ChargeLimit help with this?

**Comment by deqw** [comment]
Thanks for looking into this. Is it an option to also set StorageRemoteCtrl_ChargeLimit to a high-value?

"3 – Charge from PV+AC according to the max battery power.
Charge from both PV and AC with priority on PV power.
**If PV production is lower than StorageRemoteCtrl_ChargeLimit, the battery will be charged from AC up to NominalActivePow-erLimit.** In this case AC power = StorageRemoteCtrl_ChargeLimit- PVpower.
If PV power is larger than StorageRemoteCtrl_ChargeLimit the excess PV power will be directed to the AC up to the Nominal-ActivePowerLimit beyond which the PV is curtailed."

**Reply by WillCodeForCats** [reply]
I don't have batteries, but you can try that.

**Reply by deqw** [reply]
If I could. However I'm only using your solaredge-modbus-multi implementation to get and set these kind of modbus items for my Solaredge inverter. As far as I can see this property is not available.

**Reply by WillCodeForCats** [reply]
This integration supports all storage control properties available over modbus.

**Reply by jovana** [reply]
> "3 – Charge from PV+AC according to the max battery power.

Is this an option, number 3? Can you share the documentation where you find this?

**Reply by deqw** [reply]
@jovana Found it here (page 17): [https://www.photovoltaikforum.com/core/attachment/88445-power-control-open-protocol-for-solaredge-inverters-pdf/](url)

Can't find this property anyway in this integration.

**Reply by jovana** [reply]
Ah okay! This is the STORAGE_MODE setting. But what I make of it: If you set the STORAGE_CHARGE_LIMIT to a higher value then your PV production, then it gets power from AC. Not tested, but this is what I make of it.

**Reply by WillCodeForCats** [reply]
Start with the wiki, not docs found on other websites:

https://github.com/WillCodeForCats/solaredge-modbus-multi/wiki/Power-Control-Options

You can see a sample of my simulator dashboard which has all the controls:

https://github.com/WillCodeForCats/solaredge-modbus-multi/wiki/YAML-Sample-Configurations

You will not find an option called "3 Charge from PV+AC" here, it's called "Charge from Solar Power and Grid". If you're going to use other docs then you need to understand what they mean and not look for literal matches. I assure you the options are here and if you can't find then then you aren't looking in the right place. Post screenshots of what you do have if you can't figure it out.

**Reply by deqw** [reply]
Ok, so StorageRemoteCtrl_ChargeLimit == STORAGE_CHARGE_LIMIT
Thanks for clearing this up to me.

**Comment by jovana** [comment]
> Ok, so StorageRemoteCtrl_ChargeLimit == STORAGE_CHARGE_LIMIT Thanks for clearing this up to me.

It is the STORAGE_COMMAND_MODE set to option 3 (Charge from PV+AC). But I think you need also to play on the STORAGE_CHARGE_LIMIT. Set it higher then your panels will produce.

Also what you can try, is the setting in Charge Policy = Percent of Production. What will happen if you put it on 1%? Is the rest charge from AC?

**Reply by deqw** [reply]
Yes that is what the idea was two days ago, only had to relate the property mentioned in the documentation with the property in the integration. But, thanks again, will test this soon.

In the same document:
Percent of Production - allows AC charging with a % of system production year to date limit (needed for meeting ITC regulation in the US)

**Reply by jovana** [reply]
Very odd there is no overwrite to disable PV power completely. On the other hand, we have a very unique situation because we need this of the negative prices (sometimes).

**Reply by jovana** [reply]
I have also asked someone I know at Eneco (dutch energy provider) where they have access to the SolarEdge dev team. I hope they can sort something out. 🤞

**Reply by WillCodeForCats** [reply]
Wouldn't you just set a site limit of 0 for production control or export control under the site limit controls?

**Reply by jovana** [reply]
Yes, but for some reason the kept enable to charge the battery. If the battery is 100% the PV production is indeed to 0.

**Reply by deqw** [reply]
I added the automation to handle this last week but now I'm looking to add setting sensor.solaredge_i1_active_power_limit to the automation. Eg. set it to 0 when negative pricing and back to 100 when pricing is above zero. This should limit PV production at all I think, but it looks like the configuration setting for this property is missing. Can you help/advise? Thanks!

**Reply by WillCodeForCats** [reply]
Set `select.solaredge_i1_limit_control_mode` to `Production Control` and then set `number.solaredge_i1_site_limit` to 0.

**Reply by deqw** [reply]
Thanks @WillCodeForCats , that works.

**Reply by deqw** [reply]
@WillCodeForCats Can you say why you choose the Active Power Limit cannot be set?

**Reply by WillCodeForCats** [reply]
Production control already did that, so it was never talked about.

**Reply by WillCodeForCats** [reply]
[Release v2.4.4-pre.2](https://github.com/WillCodeForCats/solaredge-modbus-multi/releases/tag/v2.4.4-pre.2) adds Active Power Limit and CosPhi settings.

**Reply by deqw** [reply]
Thanks. Will adapt the automation when this is released.

**Reply by WillCodeForCats** [reply]
If you can test it that would be best, I'm not aware that anyone else has wanted that option.

**Reply by deqw** [reply]
I installed the v2.4.4 release and tested the Active Power Limit and it works as documented: This register controls the active power limit of the inverter dynamically. It is set as the percentage of the inverter’s maximum power. For example: when setting the register of SE5000 to 20, it will limit the inverter to 1000W which is 20% of 5000W.

Setting Active Power Limit to zero results in taking only power from the grid which can be used when having negative pricing. There is no need to change the Storage Control Mode of the inverter so I can leave it to Maximise Self Consumption while applying Active Power Limit.

Edit: when Maximize Self Consumption and a Backup battery is in use prevent using battery input to the house by setting Backup Reserve to 100% while having negative pricing (and change back to whatever you like when the tariff comes above zero again).

@WillCodeForCats Thanks for adding this setting!

BTW: I noticed that Grid Control does not maintain its setting (when setting Off it reverts to On after refresh), this was already the case with the version before v2.4.4. so is no specific result of this update.
